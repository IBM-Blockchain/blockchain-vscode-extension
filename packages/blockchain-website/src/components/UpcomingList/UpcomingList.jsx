import React, { Component } from 'react'
import Axios from 'axios'
import IssuesList from '../IssuesList/IssuesList'

const githubProject = 'IBM-Blockchain/blockchain-vscode-extension';

//Github URLs
const issuesNextApiUrl = 'https://api.github.com/repos/'+githubProject+'/issues?labels=next&state=closed&per_page=100'
const allMilestones = 'https://api.github.com/repos/'+githubProject+'/milestones'

class UpcomingList extends Component {

    constructor() {
        super();
        this.state = {
            issues: undefined,
            error: undefined
        }

        this.refreshIssues = this.refreshIssues.bind(this);
    }

    async componentDidMount(){
        const milestoneResponse = await Axios.get(allMilestones);
        if (milestoneResponse) {
            const newestMilestone = milestoneResponse.data.reduce(function(a,b) {
                return (a.due_on > b.due_on) ? a : b;
            });
            this.setState({
                newestMilestone: newestMilestone
            });
        } else {
            this.setState({
                error : 'Cannot load from Github, sorry.'
            });        }

        await this.refreshIssues()
    }

    async refreshIssues(){
        const response = await Axios.get(issuesNextApiUrl);
        if (response) {
            this.setState({
                issues: response.data
            });
        } else {
            this.setState({
                error : 'Cannot load from Github, sorry.'
            });
        }
    }

    render() {
        let latestDueDate = '';
        let milestoneNumber = '';
        let issuesThisMilestone;
        if (this.state.newestMilestone) {
            const dt = new Date(this.state.newestMilestone.due_on);

            // Set expected release date to milestone due_date minus days (thursday before)
            dt.setDate(dt.getDate()-4);

            latestDueDate = dt.toLocaleDateString();
            milestoneNumber = this.state.newestMilestone.number;
        }

        if (this.state.issues || this.state.error) {
            if (this.state.issues && this.state.newestMilestone) {
                const milestoneTitle = this.state.newestMilestone.title
                issuesThisMilestone = this.state.issues.filter(function(issue) {
                    if(issue.milestone) {
                        if (issue.milestone.title === milestoneTitle) {
                            return issue;
                        }
                    }
                });

            }
    
            return (
                <div>
                    <IssuesList 
                        key='issues' 
                        issues={issuesThisMilestone}
                        error={this.state.error}
                        refreshIssues={this.refreshIssues}
                        title={'Upcoming release: expected on ' + latestDueDate}
                        issueLabel='next'
                        newestMilestone= {this.state.newestMilestone}
                        milestoneNumber={milestoneNumber}>
                    </IssuesList>
                </div>
            );
        } else {
            return <></>;
        }
    }
}

export default UpcomingList