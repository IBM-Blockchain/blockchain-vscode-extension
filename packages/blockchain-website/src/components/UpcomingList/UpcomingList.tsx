import React, { Component } from 'react';
import Axios, { AxiosResponse } from 'axios';
import IssuesList from '../IssuesList/IssuesList';

const githubProject: string = 'IBM-Blockchain/blockchain-vscode-extension';

// Github URLs
const issuesNextApiUrl: string = 'https://api.github.com/repos/' + githubProject + '/issues?labels=next&state=closed&per_page=100';
const allMilestones: string = 'https://api.github.com/repos/' + githubProject + '/milestones';

interface UpcomingState {
    issues: any;
    error: string | undefined;
    newestMilestone: any;
}

class UpcomingList extends Component<{}, UpcomingState> {

    constructor(props: Readonly<UpcomingState>) {
        super(props);
        this.state = {
            issues: undefined,
            error: undefined,
            newestMilestone: undefined
        };

        this.refreshIssues = this.refreshIssues.bind(this);
    }

    async componentDidMount(): Promise <void> {
        try {
            const milestoneResponse: AxiosResponse = await Axios.get(allMilestones);
            const currentMilestone: number = milestoneResponse.data.reduce((a: any, b: any) => {
                return (a.due_on > b.due_on) ? a : b;
            });
            this.setState({
                newestMilestone: currentMilestone
            });
        } catch (error) {
            this.setState({
                error : 'Cannot load from Github, sorry.'
            });
        }

        await this.refreshIssues();
    }

    async refreshIssues(): Promise<void> {
        try {
            const response: AxiosResponse = await Axios.get(issuesNextApiUrl);
            this.setState({
                issues: response.data
            });
        } catch (error) {
            this.setState({
                error : 'Cannot load from Github, sorry.'
            });
        }
    }

    render(): JSX.Element {
        let latestDueDate: string = '';
        let milestoneNumber: string = '';
        let issuesThisMilestone: any;
        if (this.state.newestMilestone) {
            const dt: Date = new Date(this.state.newestMilestone.due_on);

            // Set expected release date to milestone due_date minus days (thursday before)
            dt.setDate(dt.getDate() - 4);

            latestDueDate = dt.toLocaleDateString();
            milestoneNumber = this.state.newestMilestone.number;
        }

        if (this.state.issues || this.state.error) {
            if (this.state.issues && this.state.newestMilestone) {
                const milestoneTitle: string = this.state.newestMilestone.title;

                issuesThisMilestone = this.state.issues.filter((issue: any) => {
                    if (issue.milestone) {
                        if (issue.milestone.title === milestoneTitle) {
                            return issue;
                        }
                    }
                });
            }
            let title: string = '';
            let newMilestone: any;

            if (this.state.error) {
                title = 'Fixed in upcoming release';
                newMilestone = 'someMilestone';
            } else {
                title = 'Fixed in upcoming release (expected on ' + latestDueDate + ')';
                newMilestone = this.state.newestMilestone;
            }

            return (
                <div>
                    <IssuesList
                        key='issues'
                        issues={issuesThisMilestone}
                        error={this.state.error}
                        title={title}
                        issueLabel='next'
                        newestMilestone= {newMilestone}
                        milestoneNumber={milestoneNumber}>
                    </IssuesList>
                </div>
            );
        } else {
            return <></>;
        }
    }
}

export default UpcomingList;
