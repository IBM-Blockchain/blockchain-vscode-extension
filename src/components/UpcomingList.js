import React, { Component } from 'react'
import $ from 'jquery'
import IssuesList from './IssuesList'

var githubProject = 'IBM-Blockchain/blockchain-vscode-extension';

//Github URLs
var issuesNextApiUrl = 'https://api.github.com/repos/'+githubProject+'/issues?labels=next&state=closed&per_page=100'
var allMilestones = 'https://api.github.com/repos/'+githubProject+'/milestones'

class UpcomingList extends Component {

    constructor() {
        super();
        this.state = {
            issues: undefined,
            error: undefined,
            refreshing: false
        }

        this.refreshIssues = this.refreshIssues.bind(this);
    }

    componentDidMount(){
        this.refreshIssues()
        // this.checkerId = setInterval(this.refreshIssues,10000);
    }

    componentWillUnmount(){
        // clearInterval(this.refreshIssues);
    }

    getMilestones() {
        
    }

    refreshIssues(e){

        if (e !== undefined) {
            e.preventDefault();
        }

        this.setState({
            refreshing : true
        });

        var updateIssues = function(data){
            this.setState({
                issues : data
            });
            //we add a timeout to give some UI feedback on the loading even if it is instantaneous
            setTimeout(function(){
                this.setState({
                    refreshing : false
                })
            }.bind(this),200);
        }.bind(this)

        $.get({
            url : allMilestones,
            success : function(data){
                let newestMilestone;

                newestMilestone = data.reduce(function(a, b) {
                    return (a.due_on > b.due_on) ? a : b;
                })

                this.setState({
                    newestMilestone: newestMilestone
                });

            }.bind(this),
            error : function(){
                //we add a timeout to give some UI feedback on the loading even if it is instantaneous
                setTimeout(function(){this.setState({refreshing : false})}.bind(this),200)
                this.setState({error : 'Cannot load from Github, sorry.'})}.bind(this)
        })

        $.get({
            url : issuesNextApiUrl,
            success : updateIssues,
            error : function(){
                //we add a timeout to give some UI feedback on the loading even if it is instantaneous
                setTimeout(function(){
                    this.setState({
                        refreshing : false
                    });
                }.bind(this),200);
                this.setState({
                    error : 'Cannot load incidents from Github, sorry.'
                });
            }.bind(this)
        });
    }

    getInitialState(){
        return {
            refreshing : false
        }
    }
    render() {
        let latestDueDate = '';
        let milestoneNumber = '';
        if (this.state.newestMilestone) {
            const dt = new Date(this.state.newestMilestone.due_on);

            // Set expected release date to milestone due_date minus days (thursday before)
            dt.setDate(dt.getDate()-4);

            latestDueDate = dt.toLocaleDateString();
            milestoneNumber = this.state.newestMilestone.number;
        }
        if (this.state.issues || this.state) {
            return (
                <div>
                    <IssuesList 
                        key='issues' issues={this.state.issues}
                        refreshing={this.state.refreshing}
                        error={this.state.error}
                        refreshIssues={this.refreshIssues}
                        title={'Upcoming release: expected on ' + latestDueDate}
                        issueLabel='next'
                        newestMilestone= {this.state.newestMilestone}
                        milestoneNumber={milestoneNumber}>
                    </IssuesList>
                </div>
            );
        } else {
            return <></>;
        }
    }
}

export default UpcomingList