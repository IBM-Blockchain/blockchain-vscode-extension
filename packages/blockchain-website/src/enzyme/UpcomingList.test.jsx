import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import chai from 'chai';
import sinon from 'sinon';
import Axios from 'axios';
import sinonChai from 'sinon-chai';
import UpcomingList from '../components/UpcomingList/UpcomingList.jsx';

jest.useFakeTimers();
chai.should();
chai.use(sinonChai);

const githubProject = 'IBM-Blockchain/blockchain-vscode-extension';

const issuesNextApiUrl = 'https://api.github.com/repos/'+githubProject+'/issues?labels=next&state=closed&per_page=100'
const allMilestones = 'https://api.github.com/repos/'+githubProject+'/milestones'

describe('UpcomingList', () => {
    let mySandBox;
    let axiosGetStub;

    beforeEach(async() => {
        mySandBox = sinon.createSandbox();
        axiosGetStub = mySandBox.stub(Axios, 'get');
        axiosGetStub.onFirstCall().returns({data: [
            {
                url: 'milestoneOneURL',
                title: 'Work Items',
                due_on: null
            },
            {
                url: 'anotherMilestoneURL',
                title: 'Sprint x',
                due_on: '2020-03-23T07:00:00Z'
            },
            {
                url: 'moreMilestones',
                title: 'Sprint y',
                due_on: '2020-03-20T07:00:00Z'
            }
        ]});
        axiosGetStub.onSecondCall().returns({data: [
            {
                url: 'someURL',
                repositoryurl: 'someOtherURL',
                number: 1932,
                title: 'Some completed issue',
                state: 'closed',
                labels: [{
                    id : 'someLabelID',
                    name: 'bug',
                    description: 'someLabelDescription'
                }],
                milestone: {
                    number: 'milestoneNumber',
                    id: 'someMilestoneID',
                    title: 'Sprint x'
                }
            },
            {
                url: 'someURL1',
                repositoryurl: 'someOtherURL1',
                number: 1922,
                title: 'Some other completed issue',
                state: 'closed',
                labels: [{
                    id : 'someOtherLabelID',
                    name: 'enhancement',
                    description: 'someOtherLabelDescription'
                }],
                milestone: {
                    number: 'milestoneNumber',
                    id: 'someMilestoneID',
                    title: 'Sprint y'
                }
            }
        ]});
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    it('should render the expected snapshot', async() => {
        const component = renderer
            .create(<UpcomingList />)
            .toJSON(); 
        expect(component).toMatchSnapshot();
    });

    it('should get the newest milestone with a successful api call', async() => {
        const component = await mount(<UpcomingList/>);
        axiosGetStub.should.have.been.calledTwice;
        component.state().newestMilestone.should.deep.equal({
            url: 'anotherMilestoneURL',
            title: 'Sprint x',
            due_on: '2020-03-23T07:00:00Z'
        });
    });

    it('should show error when there is an unsuccessful milestone api call', async() => {
        axiosGetStub.onFirstCall().returns(undefined);
        const component = await mount(<UpcomingList/>);
        axiosGetStub.should.have.been.calledTwice;
        component.state().error.should.equal('Cannot load from Github, sorry.');
    });

    it('should successfully make an API call to GitHub', async () => {
        await mount(<UpcomingList/>);
        axiosGetStub.should.have.been.calledTwice;
        axiosGetStub.getCall(0).should.have.been.calledWith(allMilestones);
        axiosGetStub.getCall(1).should.have.been.calledWith(issuesNextApiUrl);
    });

    it('should get latest completed issues with a successful api call', async() => {   
        const issues = [
            {
                url: 'someURL',
                repositoryurl: 'someOtherURL',
                number: 1932,
                title: 'Some completed issue',
                state: 'closed',
                labels: [{
                    id : 'someLabelID',
                    name: 'bug',
                    description: 'someLabelDescription'
                }],
                milestone: {
                    number: 'milestoneNumber',
                    id: 'someMilestoneID',
                    title: 'Sprint x'
                }
            },
            {
                url: 'someURL1',
                repositoryurl: 'someOtherURL1',
                number: 1922,
                title: 'Some other completed issue',
                state: 'closed',
                labels: [{
                    id : 'someOtherLabelID',
                    name: 'enhancement',
                    description: 'someOtherLabelDescription'
                }]
            }];  
        const component = await mount(<UpcomingList/>);
        component.setState({
            issues: issues,
            error: undefined
        })
        axiosGetStub.should.have.been.calledTwice;
        component.text().includes('Some completed issue').should.be.true;
        component.state().issues.should.deep.equal(issues);
    });

    it('should show error when there is an unsuccessful api calls', async() => {
        axiosGetStub.onFirstCall().returns(undefined);
        axiosGetStub.onSecondCall().returns(undefined);
        const component = await mount(<UpcomingList/>);
        component.state().error.should.equal('Cannot load from Github, sorry.');
        axiosGetStub.should.have.been.calledTwice;
    });
});
