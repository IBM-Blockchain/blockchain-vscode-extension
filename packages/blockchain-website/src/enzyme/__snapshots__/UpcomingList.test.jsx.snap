// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpcomingList should render the expected snapshot 1`] = `null`;
